### PR TITLE
feat: custom searcher doc

### PR DESCRIPTION
**Pull request by mpkouznetsov**
## Description

https://determinedai.atlassian.net/browse/MLG-54

- Update docstrings
- Include public classes in Custom Searcher Reference
- Add a section on custom search methods to hyperparameter tuning chapter

Note: after building this branch you can run `make live` in the docs directory to start a docs server that will listen at `localhost:1234`

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+| :doc:`/reference/reference-searcher/custom-searcher-reference` |
++----------------------------------------------------------------+
+
+Determined supports defining your own hyperparameter search algorithms and provides search runner
+utilities for executing them.
+
+Remember that a Determined experiment is a set of trials, each corresponding to a point in the
+hyperparameter space.
+
+To implement a custom hyperparameter tuning algorithm, subclass
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:19]: How about: 

To implement a custom hyperparameter tuning algorithm, subclass
:class:`~determined.searcher.SearchMethod`, overriding its event handler methods.
Note that overriding :meth:`~determined.searcher.SearchMethod.save_method_state` and
:meth:`~determined.searcher.SearchMethod.load_method_state` is not mandatory,
however, it is **recommended** as it enables fault tolerance.

To run the custom HP tuning algorithm, you can use:

- :class:`~determined.searcher.LocalSearchRunner` that executes on your local machine,
- :class:`~determined.searcher.RemoteSearchRunner` that runs on a Determined cluster.

SearchRunner executes the custom HP tuning algorithm and starts a multi-trial experiment
on a Determined cluster.

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+   import uuid
+   from pathlib import Path
+   from typing import List
+
+   class MySearchMethod(searcher.SearchMethod):
+       def __init__(self, ...) -> None:
+           super().__init__()
+           # initialize any state you would need
+
+       def initial_operations(self) -> List[searcher.Operation]:
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:55]: I think we should list available Operations or provide a link to a reference page.
Copied from examples:
```
Currently, we support the following operations:
  -> Create - starts a new trial with a unique trial id and a set of hyperparameters,
  -> ValidateAfter - sets number of steps (i.e., batches or epochs) after which a validation is run for a trial
                     with a given id,
  -> Close - closes a trial with a given id,
  -> Shutdown - closes the experiment.
```

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+       def progress(self) -> float:
+           # report experiment progress as a value between 0.0 and 1.0
+           # the Web UI will display a corresponding progress bar
+           return 0.0
+
+       def on_trial_exited_early(self) -> List[searcher.Operation]:
+           # update internal state, reflecting early trial exit
+           # return operations (e.g., create a trial with a different
+           # combination of hyperparameters
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:91]: )

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+           return []
+
+       def save_method_state(self, path: Path) -> None:
+           # save any useful state to a file you create in directory path
+           checkpoint_path = path.joinpath("method_state.json")
+           with checkpoint_path.open("w") as f:
+               # populate a dictionary or another serializable data structure
+               # with the internal state data
+               # you can use any serialization format (not just json)
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:100]: straight to the point, great!

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+           with checkpoint_path.open("r") as f:
+               state = json.load(f)
+               # initialize internal state from the deserialized data structure
+
+***********************************
+ Run Hyperparameter Search Locally
+***********************************
+
+A script performing hyperparameter tuning using :class:`~determined.searcher.LocalSearchRunner` may
+look like the following:
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:115]: Can we provide link to the example in the docs? I think I would be useful

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+       # SearchMethod.save_method_state() and SearchMethod.load_method_state() to ensure
+       # correct resumption of the SearchMethod.
+       searcher_dir = Path("local_search_runner/searcher_dir")
+
+       # instantiate your search method, passing the necessary parameters
+       search_method = MySearchMethod(...)
+
+       search_runner = searcher.LocalSearchRunner(search_method, searcher_dir=searcher_dir)
+
+       experiment_id = search_runner.run(model_config, context_dir=model_context_dir)
```
Comment (aciborowska) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:143]: Isn't it `model_dir` now? :)

File: harness/determined/searcher/_search_method.py [hunk 7]
```diff
@@ -76,12 +92,19 @@ def _from_bindings(
...
+    """
     @abstractmethod
     def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
         pass
 
 
 class ValidateAfter(Operation):
+    """
+    Operation signaling the trial to train until its total batches trained
+    equals the specified length
```
Comment (aciborowska) [inline harness/determined/searcher/_search_method.py:106]: Isn't it either epochs or batches (depending on what is passed as `unit` in the yaml config)?

File: harness/determined/searcher/_search_method.py [hunk 8]
```diff
@@ -148,36 +180,42 @@ def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
 
 class SearchMethod:
     """
-    Override this class to implement a custom hyperparameter search method.
+    The implementation of a custom hyperparameter tuning algorithm
```
Comment (aciborowska) [inline harness/determined/searcher/_search_method.py:183]: dot at the end

**Review by aciborowska — COMMENTED** [review]
suggestions and a few fixes/typos

File: docs/training/hyperparameter/search-methods/hp-custom.rst [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+To run the custom hyperparameter tuning algorithm, you can use:
+
+-  :class:`~determined.searcher.LocalSearchRunner` that executes on your machine,
+-  :class:`~determined.searcher.RemoteSearchRunner` that runs on a Determined cluster.
+
+Both Search Runners execute the custom hyperparameter tuning algorithm and start a multi-trial
+experiment on a Determined cluster.
+
+The following sections explain the steps to take in order to implement and use a custom
+hyperparameter search algorithm. A detailed example can be found here:
```
Comment (mpkouznetsov) [inline docs/training/hyperparameter/search-methods/hp-custom.rst:37]: This is where I decided to put the link to the example. The reader can choose to work through a detailed example _or_ read the schematic explanation.

**Review by aciborowska — APPROVED** [review]
LGTM
